jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -351,35 +330,67 @@ private ContainerConfigurationTemplate pullContainerConfigJson(
+   * @throws UnlistedPlatformInManifestListException if a cached manifest list has no manifests
+   *     matching the configured platform
    */
-  private Optional<Image> getCachedBaseImage()
+  @VisibleForTesting
+  List<Image> getCachedBaseImages()
       throws IOException, CacheCorruptedException, BadContainerConfigurationFormatException,
-          LayerCountMismatchException {
+          LayerCountMismatchException, UnlistedPlatformInManifestListException {
     ImageReference baseImage = buildContext.getBaseImageConfiguration().getImage();
     Optional<ImageMetadataTemplate> metadata =
         buildContext.getBaseImageLayersCache().retrieveMetadata(baseImage);
     if (!metadata.isPresent()) {
-      return Optional.empty();
+      return Collections.emptyList();
     }
 
+    ManifestTemplate manifestList = metadata.get().getManifestList();
     List<ManifestAndConfigTemplate> manifestsAndConfigs = metadata.get().getManifestsAndConfigs();
-    Verify.verify(manifestsAndConfigs.size() == 1);
-    ManifestTemplate manifest = Verify.verifyNotNull(manifestsAndConfigs.get(0).getManifest());
-    if (manifest instanceof V21ManifestTemplate) {
-      return Optional.of(JsonToImageTranslator.toImage((V21ManifestTemplate) manifest));
+
+    if (manifestList == null) {
+      Verify.verify(manifestsAndConfigs.size() == 1);
+      ManifestTemplate manifest = manifestsAndConfigs.get(0).getManifest();
+      if (manifest instanceof V21ManifestTemplate) {
+        return Collections.singletonList(
+            JsonToImageTranslator.toImage((V21ManifestTemplate) manifest));
+      }
+      return Collections.singletonList(
+          JsonToImageTranslator.toImage(
+              (BuildableManifestTemplate) Verify.verifyNotNull(manifest),
+              Verify.verifyNotNull(manifestsAndConfigs.get(0).getConfig())));
     }
 
-    return Optional.of(
-        JsonToImageTranslator.toImage(
-            (BuildableManifestTemplate) manifest,
-            Verify.verifyNotNull(manifestsAndConfigs.get(0).getConfig())));
+    // Manifest list cached. Identify matching platforms and check if all of them are cached.
+    ImmutableList.Builder<Image> images = ImmutableList.builder();
+    for (Platform platform : buildContext.getContainerConfiguration().getPlatforms()) {
+      String manifestDigest =
+          lookUpPlatformSpecificImageManifest((V22ManifestListTemplate) manifestList, platform);
+
+      Optional<ManifestAndConfigTemplate> manifestAndConfigFound =
+          manifestsAndConfigs
+              .stream()
+              .filter(entry -> manifestDigest.equals(entry.getManifestDigest()))
+              .findFirst();
+      if (!manifestAndConfigFound.isPresent()) {
+        return Collections.emptyList();
+      }
+
+      ManifestTemplate manifest = Verify.verifyNotNull(manifestAndConfigFound.get().getManifest());
+      ContainerConfigurationTemplate containerConfig =
+          Verify.verifyNotNull(manifestAndConfigFound.get().getConfig());
```

Comment:
I may be missing something here but what's the reasoning behind checking if these values are null?

Reply:
It's our assumption and how we designed the cache to work. If any `ManifestAndConfig` entry (which is just a pair of a manifest and a config) is saved (cached), their values should never be null; otherwise, something got terribly wrong.

Now, we then use `Verify.verifyNotNull()` to signal the NullAway check (it's a check we set up in our repo to catch potential NPE) that `manifest` and `containerConfig` are not supposed to be null. Otherwise, because `JsonToImageTranslator.toImage()` doesn't except `@Nullable` parameters, NullAway will complain that you're passing potentially null parameters.

Reply:
Oh I see! Makes sense